File: csharp/Camera.cs
using System;
using System.Numerics;

namespace ToyRayTrace
{
    public readonly struct Camera
    {
        readonly Vector3 m_Origin;
        readonly Vector3 m_LowerLeftCorner;
        readonly Vector3 m_Horizontal;
        readonly Vector3 m_Vertical;
        readonly float m_LensRadius;
        readonly Vector3 m_U, m_V, m_W; // Camera orientation

        public Camera(in Vector3 lookFrom, in Vector3 lookAt, in Vector3 vUp, float vFov, float aspect, float aperture = 0, float focusDistance = 1)
        {
            m_LensRadius = aperture / 2f;
            var theta = vFov * MathF.PI / 180f;
            var halfHeight = MathF.Tan(theta / 2f);
            var halfWidth = aspect * halfHeight;

            m_W = Vector3.Normalize(lookFrom - lookAt);
            m_U = Vector3.Normalize(Vector3.Cross(vUp, m_W));
            m_V = Vector3.Cross(m_W, m_U);

            m_Origin = lookFrom;
            m_LowerLeftCorner = m_Origin -
                m_U * halfWidth * focusDistance -
                m_V * halfHeight * focusDistance -
                m_W * focusDistance;
            m_Horizontal = m_U * 2f * halfWidth * focusDistance;
            m_Vertical = m_V * 2f * halfHeight * focusDistance;
        }

        public Ray GetRay(float s, float t, ref uint state)
        {
            var random = Rng.NextInUnitDisc(ref state) * m_LensRadius;
            var offset = m_U * random.X + m_V * random.Y;
            return new Ray(m_Origin + offset,
                m_LowerLeftCorner + m_Horizontal * s + m_Vertical * t - m_Origin - offset);
        }
    }
}

File: rust/src/raytrace/camera.rs
use super::vec3;
use super::Ray;
use super::Vec3;
use std::f32::consts;

pub struct Camera {
    origin: Vec3,
    lower_left_corner: Vec3,
    horizontal: Vec3,
    vertical: Vec3,
    u: Vec3,
    v: Vec3,
    w: Vec3,
    lens_radius: f32,
}

impl Camera {
    pub fn new(
        look_from: Vec3,
        look_at: Vec3,
        v_up: Vec3,
        v_fov: f32,
        aspect: f32,
        aperture: f32,
        focus_dist: f32,
    ) -> Self {
        let lens_radius = aperture / 2.0;
        let theta = v_fov * consts::PI / 180.0;
        let half_height = (theta / 2.0).tan();
        let half_width = aspect * half_height;

        let w = vec3::normalize(look_from - look_at);
        let u = vec3::normalize(vec3::cross(v_up, w));
        let v = vec3::cross(w, u);
        let origin = look_from;
        let lower_left_corner = origin - u * focus_dist * half_width - v * focus_dist * half_height - w * focus_dist;
        let horizontal = u * 2.0 * focus_dist * half_width;
        let vertical = v * 2.0 * focus_dist * half_height;
        Camera {
            origin,
            lower_left_corner,
            horizontal,
            vertical,
            u,
            v,
            w,
            lens_radius,
        }
    }

    pub fn get_ray(&self, s: f32, t: f32) -> Ray {
        let rd = vec3::random_in_unit_disc() * self.lens_radius;
        let offset = self.u * rd.x + self.v * rd.y;
        Ray::new(
            self.origin + offset,
            self.lower_left_corner + self.horizontal * s + self.vertical * t - self.origin - offset,
        )
    }
}
